More flexible template resolution

vkurup opened the pull request:
This implements the proposal suggested in #28.

Note that I updated the documentation to make it clear that the app-specific templates need to be formatted `{model}{activity}.html` without an underscore. E.g `smsread.html`, not `sms_read.html`. This is because we have set `template_name_suffix` to be `activity.html`, not `_activity.html` as they are in Django Vanilla Views.

dpoirier commented:
Would you care to undertake adding a templates page to the bread docs, with a section spelling out how a bread view comes up with a template name to use, and some examples of when you might use the different options?  The examples in particular would help me to get my head around why one behavior (like in this pull request) is better than another.

vkurup commented:
Sure, that's a good idea.

vkurup commented:
Ready for review. The docs still seem a little fuzzy to me, so feedback is welcome. We definitely have a lot of customization options with templates.

A separate point that I'd like feedback on: The current default app-specific location to be searched is '{model}{activity}.html' (e.g. 'smsread.html'). I think it would be better to have an underscore between them. Making that small change would make django-bread use 'bread/_read.html' as a fallback, which is a little ugly. I propose changing the code to allow both to be pretty: 'register/sms_read.html' first, and then 'bread/read.html' as a fallback.

dpoirier commented on docs/template.rst:
```diff
@@ -0,0 +1,93 @@
+
+   class MyBrowseView(BrowseView):
+       def get_context_data(self, **kwargs):
+           data = super(BreadViewMixin, self).get_context_data(**kwargs)
```
s/BreadViewMixin/BrowseView/

dpoirier commented on docs/template.rst:
```diff
@@ -0,0 +1,93 @@
+App-specific location
+^^^^^^^^^^^^^^^^^^^^^
+
+The app-specific location is described by this pattern: ``{app_name}/{model_name}{activity}.html``.
```
Could we document this pattern using the same syntax and variables as the user would use to specify their own custom template name patter?

dpoirier commented on docs/template.rst:
```diff
@@ -0,0 +1,93 @@
+2. Specify a Django setting ``BREAD['DEFAULT_TEMPLATE_NAME_PATTERN'] = 'blah'``. This will function
+   like #1 above, except for ALL bread objects on your site. You can still override this
+   customization with #1 for a specific bread object. Use this option to provide a site-wide
+   default. If, for some reason, you omit one of the action templates, then we'll fall back to the
```
I wonder if we could be even more explicit that setting DEFAULT_TEMPLATE_NAME_PATTERN to 'blah' is equivalent to setting template_name_pattern to 'blah' on every bread class that doesn't set template_name_pattern itself, and that it doesn't add another place to the list of places we look for templates?

dpoirier replied:
Another consideration - do we really need another setting? Given the way we configure Bread now, someone could easily make their own subclass of Bread that sets a default template_name_pattern and then just use that for more specific subclasses everywhere instead of subclassing Bread directly.

dpoirier commented:
>  A separate point that I'd like feedback on: The current default app-specific location to be searched is '{model}{activity}.html' (e.g. 'smsread.html'). I think it would be better to have an underscore between them. Making that small change would make django-bread use 'bread/_read.html' as a fallback, which is a little ugly. I propose changing the code to allow both to be pretty: 'register/sms_read.html' first, and then 'bread/read.html' as a fallback.

That sounds fine to me.  I thought we had an underscore in there already, frankly.

dpoirier commented on bread/__init__.py:
```diff
@@ -1 +1 @@
-VERSION = '0.1.1'
+VERSION = '0.1.2'
```
Actually I think we should leave the version alone until we do a "release", since we don't know what order we'll end up merging pulls, don't you think?

(And yes I know I bumped the version in a pull earlier. I've changed my mind about that :-) )

vkurup replied:
Makes complete sense. I assume I was just following your lead :)

dpoirier commented on bread/bread.py:
```diff
@@ -105,17 +102,19 @@ def dispatch(self, request, *args, **kwargs):
-        return (super(BreadViewMixin, self).get_template_names()
-                + ['bread/%s.html' % self.template_name_suffix])
+            return (vanilla_templates + custom_templates + default_templates)
+        return (vanilla_templates + default_templates)
```
Can we remove these parentheses?  The `return` doesn't need them, and when I first scanned the code I thought we were constructing tuples until I looked closer.

dpoirier commented on bread/bread.py:
```diff
@@ -105,17 +102,19 @@ def dispatch(self, request, *args, **kwargs):
+        default_templates = ['bread/%s.html' % self.template_name_suffix]
         if self.bread.template_name_pattern:
-            return [self.bread.template_name_pattern.format(
+            custom_templates = [self.bread.template_name_pattern.format(
```
Can we change `default_templates` and `custom_templates` to just `default_template` and `custom_template` and not have them be lists? Logically there's only ever going to be one value for each of these.

dpoirier commented on docs/templates.rst:
```diff
@@ -3,4 +3,90 @@
+Template Context
+----------------
+
+In addition to what Django Vanilla Views provides, Django Bread provides the following to all of
```
Let's add the word "context" somewhere in this sentence so it makes sense without reading the section title.

dpoirier commented on docs/templates.rst:
```diff
@@ -3,4 +3,90 @@
+We mentioned above that each View is matched to a template with the same name (BrowseView ->
+'...browse.html'). This is true for everything except the AddView. Because add and edit templates
+are so similar, the AddView connects to the 'edit.html' template. There is no 'add.html' template.
+Of course, this can be customized as described above, if you need different templates for your
```
Maybe change slightly to something like: `Of course, this can be customized using method #3 above,...`?  (I don't think you can change it by the other methods.)

vkurup replied:
There are 2 ways to customize it. Either with method # 3 above, or by setting `AddView.template_name_suffix` to `add`. (It defaults to `edit`) and then providing a `model_add.html` in the appropriate location. I'll document both.

dpoirier commented:
done re-reviewing

vkurup commented:
Ready for re-re-review. Commit 5b2228d addresses feedback. Commit aa9d02c documents an easier way to customize a specific view's template. Finally commit ee5185a implements the proposal to include an underscore between model and view in some templates, with no underscore in the default Bread templates.

dpoirier commented on bread/bread.py:
```diff
@@ -105,17 +102,26 @@ def dispatch(self, request, *args, **kwargs):
+        # Vanilla Views). If it does, then we strip the underscore to get our 'view' name.
+        # e.g. template_name_suffix '_browse' -> view 'browse'
+        suffix = self.template_name_suffix
+        view = suffix[1:] if suffix.startswith('_') else suffix
```
Or `view = self.template_name_suffix.lstrip('_')`?

vkurup replied:
much better!

dpoirier commented on docs/templates.rst:
```diff
@@ -3,4 +3,102 @@
+   the names ``app_label``, ``model``, or ``view``. An example of a valid pattern would be
+   ``'{app_label}/special_{model}_{view}.html'``.
+
+2. Create a subclass of Bread and set ``template_name_pattern`` in that subclass. Then, use that
```
I guess this isn't really another way to do it, it's just taking advantage of Python to do number 1 more efficiently, right?  Maybe this should just be a parenthetical suggestion on the end of number 1.  Or maybe I'm being too picky :-)

dpoirier commented on docs/templates.rst:
```diff
@@ -3,4 +3,102 @@
+Customization
+^^^^^^^^^^^^^
+
+There are 3 ways that you can customize this behavior even further:
```
Right now the list has 4 ways to customize the behavior.

vkurup replied:
Updated by merging customization point #2 into #1 (as you suggested) and thus making this statement correct :)

vkurup commented:
Updated.

dpoirier commented:
:+1:
